### PR TITLE
Problem (#1483): client rpc setup code is duplicated between server and c-api

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- consolidate the client-rpc setup code
- cro-clib add mock mode api when build in debug
- build cro-clib as cdylib, can be used directly by other language.
- make callback parameter in c api optional

### CODE REVIEW

File: cro-clib/Cargo.toml [hunk 1]
```diff
@@ -6,7 +6,7 @@ edition = "2018"
 
 [lib]
 name = "cro_clib"
-crate-type =["staticlib"]
+crate-type =["staticlib", "cdylib"]
```
Comment (leejw51crypto) [inline cro-clib/Cargo.toml:9]: dynamic library is intentionally removed.
because of ssl related error

File: cro-clib/chain.h [hunk 2]
```diff
@@ -19,6 +19,8 @@ typedef struct CroJsonRpc CroJsonRpc;
 
 typedef struct CroTx CroTx;
 
+typedef struct Option_ProgressCallback Option_ProgressCallback;
```
Comment (leejw51crypto) [inline cro-clib/chain.h:22]: where is Option_ProgressCallback?
opaque pointer?

Comment (leejw51crypto) [reply cro-clib/chain.h:22]: how about "OptionProgressCallback"?
removing _

Comment (yihuang) [reply cro-clib/chain.h:22]: This is auto-generated by cbindgen from `Option<ProgressCallback>`.

Comment (tomtau) [reply cro-clib/chain.h:22]: I guess the example needs updating: https://github.com/crypto-com/chain/blob/master/cro-clib/examples/rpc.c#L12 ?

Comment (tomtau) [reply cro-clib/chain.h:22]: or if one passes non-null function pointer, it'll work as intended / before?

File: cro-clib/Cargo.toml (hunk 1 above)
Comment (tomtau) [reply cro-clib/Cargo.toml:9]: @yihuang is this necessary? if it is, these instructions will need update: https://github.com/crypto-com/chain/blob/master/cro-clib/android.md as one may need to remove `cdylib` manually

File: cro-clib/chain.h (hunk 2 above)
Comment (yihuang) [reply cro-clib/chain.h:22]: > or if one passes non-null function pointer, it'll work as intended / before?

yes, the null pointer will be treated as a `None`.

File: cro-clib/Cargo.toml (hunk 1 above)
Comment (yihuang) [reply cro-clib/Cargo.toml:9]: How about moving the c wrapper for jsonrpc API into the `client-rpc-core` crate, produce `cdylib` there. `cro-clib` produce staticlib and cross-compiling as before.

Comment (tomtau) [reply cro-clib/Cargo.toml:9]: I think that works -- probably https://github.com/crypto-com/chain/blob/master/cro-clib/build.rs will need updating?

Comment (yihuang) [reply cro-clib/Cargo.toml]: Removed the `cdylib`, change the `crate-type` in `build.sh` with tool on the fly to build the shared library.

**Pull request merged**